test/Microsoft.ML.Predictor.Tests/TestIniModels.cs

```diff
@@ -504,4 +511,95 @@ public ProcessDebugInformation RunCommandLine(string commandLine, string dir)
         }
     }
 #endif
+
+    public sealed class TestIniModels : TestDataPipeBase
+    {
+        public TestIniModels(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void TestGamRegressionIni()
+        {
+            var mlContext = new MLContext(seed: 0);
+            var idv = mlContext.Data.CreateTextReader(
+                    new TextLoader.Arguments()
+                    {
+                        HasHeader = false,
+                        Column = new[]
+                        {
+                            new TextLoader.Column("Label", DataKind.R4, 0),
+                            new TextLoader.Column("Features", DataKind.R4, 1, 9)
+                        }
+                    }).Read(GetDataPath("breast-cancer.txt"));
+
+            var pipeline = mlContext.Transforms.ReplaceMissingValues("Features")
+                .Append(mlContext.Regression.Trainers.GeneralizedAdditiveModels());
+            var model = pipeline.Fit(idv);
+            var data = model.Transform(idv);
+
+            var roleMappedSchema = new RoleMappedSchema(data.Schema, false,
+                new KeyValuePair<RoleMappedSchema.ColumnRole, string>(RoleMappedSchema.ColumnRole.Feature, "Features"),
+                new KeyValuePair<RoleMappedSchema.ColumnRole, string>(RoleMappedSchema.ColumnRole.Label, "Label"));
+
+            string modelIniPath = GetOutputPath(FullTestName + "-model.ini");
+            using (Stream iniStream = File.Create(modelIniPath))
+            using (StreamWriter iniWriter = Utils.OpenWriter(iniStream))
+                model.LastTransformer.Model.SaveAsIni(iniWriter, roleMappedSchema);
+
+            var results = mlContext.Regression.Evaluate(data);
+
+            // Getting parity results from maml.exe:
+            // maml.exe ini ini=model.ini out=model_ini.zip data=breast-cancer.txt  loader=TextLoader{col=Label:R4:0 col=Features:R4:1-9} xf=NAHandleTransform{col=Features slot=- ind=-} kind=Regression
+            Assert.Equal(0.093256807643323947, results.L1);
```

Comment:
I'm not sure about this test. On one hand, it's good to make sure that it can produce the score. On the other hand, it's a new baseline test for GAMs. I'd say remove it from ML.NET.

Reply:
So, do you want me to remove both TestGamRegressionIni and TestGamBinaryClassificationIni tests?

---
In reply to: [243141824](https://github.com/dotnet/machinelearning/pull/1929#discussion_r243141824) [](ancestors = 243141824)

Reply:
We definitely need a test that checks that the new Ini file produces the same results as the GAM. It would be ideal if you could compare the the outputs with a full file compare instead of using Assert.Equal. Is it possible to do that?

---
In reply to: [244387146](https://github.com/dotnet/machinelearning/pull/1929#discussion_r244387146) [](ancestors = 244387146,243141824)

Reply:
Yes, it's better to keep these parity tests and also prevent regressions cause by refactoring in fasttree code.

The problem with ini file comparison is the same as testing baseline comparison, which needs to extract numbers in the file and compare them with some tolerance, because verbatim comparison will fail due to changes in the higher decimal places. We already have those facilities but they're fitted for train-test / CV type tasks that are common in ML. I didn't want to modify those files for this one-off ini-file comparison. Also half of this testing is done in TLC, b/c the ini parsing code is only in TLC. The baseline metrics actually come from TLC.

More importantly, the ini-format is not an open-source format and is Microsoft IP. If we want to check-in an ini file for baseline, we should have LCA clear that. 

For those reasons, I think this simple test is good enough for now. 

---
In reply to: [244822621](https://github.com/dotnet/machinelearning/pull/1929#discussion_r244822621) [](ancestors = 244822621,244387146,243141824)